From 9b2b7318df597191bf84329f5e6e74a56cb73d0c Mon Sep 17 00:00:00 2001
From: Sasha Levin <sashal@kernel.org>
Date: Thu, 30 Jul 2026 16:37:39 +0930
Subject: btrfs: disable bs > ps support if no transparent hugepage support

From: Qu Wenruo <wqu@suse.com>

[ Upstream commit 4c375ac546ea667e619ab77d34cca6edcab7c448 ]

Btrfs relies on mapping_set_folio_order_range() to set the minimal
folio order for all its data inodes, but that function will be no-op if
transparent hugepage is not enabled.

Guard the bs > ps support behind CONFIG_TRANSPARENT_HUGEPAGE, just like
all other filesystems.

Fixes: 98077f7f2180 ("btrfs: enable experimental bs > ps support")
Reviewed-by: Filipe Manana <fdmanana@suse.com>
Signed-off-by: Qu Wenruo <wqu@suse.com>
Signed-off-by: David Sterba <dsterba@suse.com>
Signed-off-by: Sasha Levin <sashal@kernel.org>
---
 fs/btrfs/Kconfig |  3 ++-
 fs/btrfs/fs.c    | 12 ++++++++++--
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/fs/btrfs/Kconfig b/fs/btrfs/Kconfig
index 4bafbf6636cd2..5f922563b6b62 100644
--- a/fs/btrfs/Kconfig
+++ b/fs/btrfs/Kconfig
@@ -115,7 +115,8 @@ config BTRFS_EXPERIMENTAL
 
 	  - extent tree v2 - complex rework of extent tracking
 
-	  - block size > page size support
+	  - block size > page size support - needs transparent huge page and
+					     non-HIGHMEM system
 
 	  - shutdown ioctl and auto-degradation support
 
diff --git a/fs/btrfs/fs.c b/fs/btrfs/fs.c
index feb0a2faa8379..bf86d8d5a05d9 100644
--- a/fs/btrfs/fs.c
+++ b/fs/btrfs/fs.c
@@ -94,9 +94,17 @@ bool __attribute_const__ btrfs_supported_blocksize(u32 blocksize)
 	 *
 	 * Considering HIGHMEM is such a pain to deal with and it's going
 	 * to be deprecated eventually, just reject HIGHMEM && bs > ps cases.
+	 *
+	 * Finally, for bs > ps cases, we need to set the minimal folio order,
+	 * which requires transparent hugepage.
 	 */
-	if (IS_ENABLED(CONFIG_HIGHMEM) && blocksize > PAGE_SIZE)
-		return false;
+	if (blocksize > PAGE_SIZE) {
+		if (IS_ENABLED(CONFIG_HIGHMEM))
+			return false;
+
+		if (!IS_ENABLED(CONFIG_TRANSPARENT_HUGEPAGE))
+			return false;
+	}
 	return true;
 #endif
 	return false;
-- 
2.53.0

